lib/util/stream_utils.js

```diff
@@ -575,17 +576,49 @@ shaka.util.StreamUtils = class {
    * @param {!Array.<shaka.extern.Variant>} variants
    * @param {boolean} usePersistentLicenses
    * @param {boolean} srcEquals
+   * @param {!Array<string>} preferredKeySystems
    * @exportDoc
    */
   static async getDecodingInfosForVariants(variants, usePersistentLicenses,
-      srcEquals) {
+      srcEquals, preferredKeySystems) {
     const gotDecodingInfo = variants.some((variant) =>
       variant.decodingInfos.length);
     if (gotDecodingInfo) {
       shaka.log.debug('Already got the variants\' decodingInfo.');
       return;
     }
 
+    // Try to get preferred key systems first to avoid unneeded calls to CDM.
+    for (const preferredKeySystem of preferredKeySystems) {
+      let keySystemSatisfied = false;
+      for (const variant of variants) {
+        /** @type {!Array.<!MediaDecodingConfiguration>} */
+        const decodingConfigs = shaka.util.StreamUtils.getDecodingConfigs_(
+            variant, usePersistentLicenses, srcEquals)
+            .filter((config) => {
+              const keySystem = config.keySystemConfiguration &&
+                config.keySystemConfiguration.keySystem;
+              return keySystem === preferredKeySystem;
+            });
+
+        // The reason we are performing this await in a loop rather than
+        // batching into a `promise.all` is performance related.
+        // https://github.com/shaka-project/shaka-player/pull/4708#discussion_r1022581178
+        for (const config of decodingConfigs) {
+          // eslint-disable-next-line no-await-in-loop
+          await shaka.util.StreamUtils.getDecodingInfosForVariant_(
+              variant, config);
+        }
+        if (variant.decodingInfos.length) {
+          keySystemSatisfied = true;
+        }
+      } // for (const variant of variants)
+      if (keySystemSatisfied) {
+        // Return if any preferred key system is already satisfied.
+        return;
+      }
+    } // for (const preferredKeySystem of preferredKeySystems)
```

Comment:
One thing that would help, I believe, is to filter the `decodingConfigs` in the loop below such that they don't contain any preferred key systems. That way, if there is a situation where a preferred key system is present but `MediaCapabilities.decodingInfo` fails when called on it, we won't make a second API call for it in the fallback case.

E.g.:
```js
const decodingConfigs = shaka.util.StreamUtils.getDecodingConfigs_(
          variant, usePersistentLicenses, srcEquals)
          .filter((config) => {
            const keySystem = config.keySystemConfiguration &&
                  config.keySystemConfiguration.keySystem;
            return !preferredKeySystems.includes(keySystem);
          });
```

Reply:
Thanks for suggestion. Proper change has been added.